File: merge-tonic-lib/tonic/core/tree.py
def state_to_tree(state):
    root_children_above = []
    children_above = [[] for i in range(len(state))]
    last_by_depth = [None] * len(state)
    for i in range(len(state)):
        line, depth, anchored_right, count, provenance = state[i]
        if not anchored_right:
            if depth == 0:
                root_children_above.append(i)
            else:
                children_above[last_by_depth[depth-1]].append(i)
        last_by_depth[depth] = i
    children_below = [[] for i in range(len(state))]
    last_by_depth = [None] * len(state)
    for i in range(len(state)-1,-1,-1):
        line, depth, anchored_right, count, provenance = state[i]
        if anchored_right:
            children_below[last_by_depth[depth-1]].append(i)
        last_by_depth[depth] = i
    for cb in children_below:
        cb.reverse()
    return (
        None,
        -1,
        [],
        [],
        [pull_out_tree(i, state, children_above, children_below) for i in root_children_above],
        -1,
    )

def pull_out_tree(pos, state, children_above, children_below):
    line, depth, archored_right, count, provenance = state[pos]
    return (
        line,
        count,
        provenance,
        [pull_out_tree(x, state, children_above, children_below) for x in children_below[pos]],
        [pull_out_tree(x, state, children_above, children_below) for x in children_above[pos]],
        depth,
    )

# format of tree is (line, count, provenance, [low_trees], [high_trees], depth)
# line is None for the root
# lines in output are of format (line, depth, anchored_right, count, provenance, on_left, on_right)
# state format is [(line, depth, anchored_right, count, provenance)]
def merge_trees(output, tree1, tree2, anchored_right):
    line1, count1, provenance1, lowtrees1, hightrees1, depth1 = tree1
    line2, count2, provenance2, lowtrees2, hightrees2, depth2 = tree2
    assert line1 == line2
    assert depth1 == depth2
    merge_tree_lists(output, lowtrees1, lowtrees2, True)
    if line1 is not None:
        if count1 > count2:
            provenance = provenance1
        elif count2 > count1:
            provenance = provenance2
        else:
            p1 = "|".join(provenance1)
            p2 = "|".join(provenance2)
            provenance = provenance1 if p1 <= p2 else provenance2
        output.append((line1, depth1, anchored_right, max(count1, count2), provenance, count1 % 2, count2 % 2))
    merge_tree_lists(output, hightrees1, hightrees2, False)

def merge_tree_lists(output, left_trees, right_trees, anchored_right):
    pos1 = 0
    pos2 = 0
    while pos1 < len(left_trees) or pos2 < len(right_trees):
        if pos2 == len(right_trees):
            insert_tree(output, left_trees[pos1], False, anchored_right)
            pos1 += 1
        elif pos1 == len(left_trees):
            insert_tree(output, right_trees[pos2], True, anchored_right)
            pos2 += 1
        elif left_trees[pos1][0] == right_trees[pos2][0]:
            merge_trees(output, left_trees[pos1], right_trees[pos2], anchored_right)
            pos1 += 1
            pos2 += 1
        elif left_trees[pos1][0] < right_trees[pos2][0]:
            insert_tree(output, left_trees[pos1], False, anchored_right)
            pos1 += 1
        else:
            insert_tree(output, right_trees[pos2], True, anchored_right)
            pos2 += 1

def insert_tree(output, tree, from_right, anchored_right):
    line, count, provenance, lowtrees, hightrees, depth = tree
    for new_tree in lowtrees:
        insert_tree(output, new_tree, from_right, True)
    output.append((line, depth, anchored_right, count, provenance, not from_right, from_right))
    for new_tree in hightrees:
        insert_tree(output, new_tree, from_right, False)

File: packages/tonic-core/src/tree.ts
import type { StateRow } from "./state";

/** Tree node: line, count, lowTrees, highTrees, depth; line null for root */
export type Tree = [string | null, number, string[], Tree[], Tree[], number];

export function stateToTree(state: StateRow[]): Tree {
  const rootChildrenAbove: number[] = [];
  const childrenAbove: number[][] = state.map(() => []);
  const lastByDepth: (number | null)[] = state.map(() => null);
  for (let i = 0; i < state.length; i++) {
    const [line, depth, anchored_right] = state[i]!;
    if (!anchored_right) {
      if (depth === 0) {
        rootChildrenAbove.push(i);
      } else {
        const parent = lastByDepth[depth - 1]!;
        childrenAbove[parent]!.push(i);
      }
    }
    lastByDepth[depth] = i;
  }

  const childrenBelow: number[][] = state.map(() => []);
  lastByDepth.fill(null);
  for (let i = state.length - 1; i >= 0; i--) {
    const [, depth, anchored_right] = state[i]!;
    if (anchored_right) {
      const parent = lastByDepth[depth - 1]!;
      childrenBelow[parent]!.push(i);
    }
    lastByDepth[depth] = i;
  }
  for (const cb of childrenBelow) {
    cb.reverse();
  }

  return [
    null,
    -1,
    [],
    [],
    rootChildrenAbove.map((i) => pullOutTree(i, state, childrenAbove, childrenBelow)),
    -1,
  ];
}

function pullOutTree(
  pos: number,
  state: StateRow[],
  childrenAbove: number[][],
  childrenBelow: number[][],
): Tree {
  const [line, , , count, provenance] = state[pos]!;
  const depth = state[pos]![1];
  return [
    line,
    count,
    provenance,
    childrenBelow[pos]!.map((x) => pullOutTree(x, state, childrenAbove, childrenBelow)),
    childrenAbove[pos]!.map((x) => pullOutTree(x, state, childrenAbove, childrenBelow)),
    depth,
  ];
}

/** (line, depth, anchored_right, in_child, on_left, on_right) status line from merge */
export type StatusLine = [string, number, boolean, number, string[], boolean, boolean];

export function mergeTrees(output: StatusLine[], tree1: Tree, tree2: Tree, anchored_right: boolean): void {
  const [line1, count1, provenance1, lowtrees1, hightrees1, depth1] = tree1;
  const [line2, count2, provenance2, lowtrees2, hightrees2, depth2] = tree2;
  if (line1 !== line2) {
    throw new Error("mergeTrees line mismatch");
  }
  if (depth1 !== depth2) {
    throw new Error("mergeTrees depth mismatch");
  }
  mergeTreeLists(output, lowtrees1, lowtrees2, true);
  if (line1 !== null) {
    let provenance = provenance1;
    if (count2 > count1) {
      provenance = provenance2;
    } else if (count1 === count2) {
      provenance = provenance1.join("|") <= provenance2.join("|") ? provenance1 : provenance2;
    }
    output.push([
      line1,
      depth1,
      anchored_right,
      Math.max(count1, count2),
      provenance,
      Boolean(count1 % 2),
      Boolean(count2 % 2),
    ]);
  }
  mergeTreeLists(output, hightrees1, hightrees2, false);
}

function mergeTreeLists(output: StatusLine[], leftTrees: Tree[], rightTrees: Tree[], anchored_right: boolean): void {
  let pos1 = 0;
  let pos2 = 0;
  while (pos1 < leftTrees.length || pos2 < rightTrees.length) {
    if (pos2 === rightTrees.length) {
      insertTree(output, leftTrees[pos1]!, false, anchored_right);
      pos1 += 1;
    } else if (pos1 === leftTrees.length) {
      insertTree(output, rightTrees[pos2]!, true, anchored_right);
      pos2 += 1;
    } else {
      const l = leftTrees[pos1]!;
      const r = rightTrees[pos2]!;
      const l0 = l[0]!;
      const r0 = r[0]!;
      if (l0 === r0) {
        mergeTrees(output, l, r, anchored_right);
        pos1 += 1;
        pos2 += 1;
      } else if (l0 < r0) {
        insertTree(output, l, false, anchored_right);
        pos1 += 1;
      } else {
        insertTree(output, r, true, anchored_right);
        pos2 += 1;
      }
    }
  }
}

function insertTree(output: StatusLine[], tree: Tree, fromRight: boolean, anchored_right: boolean): void {
  const [line, count, provenance, lowtrees, hightrees, depth] = tree;
  for (const newTree of lowtrees) {
    insertTree(output, newTree, fromRight, true);
  }
  output.push([line!, depth, anchored_right, count, provenance, !fromRight, fromRight]);
  for (const newTree of hightrees) {
    insertTree(output, newTree, fromRight, false);
  }
}
